ldms/src/sampler/llnl/lustre_client.c

```diff
@@ -0,0 +1,641 @@
+        }
+
+        /* copy the new_llite_tree into place over the global llite_tree */
+        memcpy(&llite_tree, &new_llite_tree, sizeof(struct rbt));
+
+        return;
+}
+
+static void llites_sample()
+{
+        struct rbn *rbn;
+
+        /* walk tree of known LLITEs */
+        RBT_FOREACH(rbn, &llite_tree) {
+                struct llite_data *llite;
+                llite = container_of(rbn, struct llite_data, llite_tree_node);
+                llite_general_sample(llite->name, llite->stats_path,
+                                   llite->general_metric_set);
+        }
+}
+
+static int sample(struct ldmsd_sampler *self)
+{
+        log_fn(LDMSD_LDEBUG, SAMP" sample() called\n");
+        if (llite_general_schema_is_initialized() < 0) {
+                if (llite_general_schema_init(schema_name) < 0) {
+                        log_fn(LDMSD_LERROR, SAMP" general schema create failed\n");
+                        return ENOMEM;
+                }
+        }
+
+        if ((time(NULL)-last_refresh) >= refresh_interval) {
+                last_refresh = time(NULL);
+                llites_refresh();
+        }
+
+        llites_sample();
+
+        return 0;
+}
+
+static void term(struct ldmsd_plugin *self)
+{
+        log_fn(LDMSD_LDEBUG, SAMP" term() called\n");
+        llites_destroy();
+        llite_general_schema_fini();
+}
+
+static ldms_set_t get_set(struct ldmsd_sampler *self)
+{
+	return NULL;
+}
+
+static const char *usage(struct ldmsd_plugin *self)
+{
+        log_fn(LDMSD_LDEBUG, SAMP" usage() called\n");
+	return  "config name=" SAMP " [producer=<prod_name>] [instance=<inst_name>] [component_id=<compid>] [schema=<sname>] [rescan_sec=<rsec>] [with_jobid=<jid>]\n"
+		"    <prod_name>  The producer name (default: the hostname)\n"
+		"    <inst_name>  The instance name (ignored, dynamically generated)\n"
+		"    <compid>     Optional unique number identifier (default: zero)\n"
+		"    <sname>      Optional schema name\n"
+                "    <rsec>       Interval in seconds between rescanning for lustre client mounts (default: 30s)\n"
+		LJI_DESC;
+	return  "config name=" SAMP;
+}
+
+static int config(struct ldmsd_plugin *self,
+                  struct attr_value_list *kwl, struct attr_value_list *avl)
+{
+        char *value;
+
+        log_fn(LDMSD_LDEBUG, SAMP" config() called\n");
+
+        value = av_value(avl, "producer");
+        if (value != NULL) {
+                snprintf(producer_name, sizeof(producer_name),
+                         "%s", value);
+        }
+
+	value = av_value(avl, "component_id");
+	if (value != NULL) {
+		component_id = (uint64_t)atoi(value);
+        }
+
+	value = av_value(avl, "instance");
+	if (value != NULL) {
+		log_fn(LDMSD_LWARNING, SAMP ": ignoring option \"instance=%s\"\n",
+                       value);
+	}
+
+	value = av_value(avl, "schema");
+	if (value != NULL) {
+                snprintf(schema_name, sizeof(schema_name), "%s", value);
+	}
+
```

Comment:
llite_general_schema_init should be called here and the function return an error if it fails. them sample() is simpler.

Reply:
I would like to see config() accept option blacklist=file_name which is a file containing a single column of metric names. Any metric name appearing in the file is omitted from the schema (and in the metric name index lookup, retain a negative number and ignore them at sample time when found in stats file). But again, I'm fine with not adding this feature in the initial version of the plugin.

Reply:
This will lead to a schema with the same name having different values. As it is, the contents are consistent. As it is, the schema seems short and sweet. Why mess with it?

Reply:
With the other requested feature (blacklist capability) the schema will not be all identical, so need to be nameable.

Reply:
I'm not really planning to add blacklists at this point.

Reply:
If we get the rest of the nits sorted, I will contribute blacklist in a subsequent patch. it's good to have schema name from the start handled as expected for all other plugins.